system/src/system_power_manager.cpp

```diff
@@ -552,52 +557,188 @@ void PowerManager::handleStateChange(battery_state_t from, battery_state_t to, b
+    DBG_PWR("Rstore normal threshold: 100mV");
+    PMIC power(true);
+    power.setRechargeThreshold(PMIC_NORMAL_RECHARGE_THRESHOLD); // 100mV
+    return;
+  }
+
+  if (millis() - batMonitorTimeStamp_ >= battMonitorPeriod_) {
+    batMonitorTimeStamp_ = millis();
+    if (config_.flags & HAL_POWER_CHARGE_STATE_DISABLE) {
+      deduceBatteryStateChargeDisabled();
+    } else {
+      deduceBatteryStateChargeEnabled();
+    }
   }
 }
 
-void PowerManager::checkWatchdog() {
-  if (g_batteryState == BATTERY_STATE_DISCONNECTED &&
-      ((millis() - chargingDisabledTimestamp_) >= DEFAULT_WATCHDOG_TIMEOUT)) {
-    // Re-enable charging, do not run DPDM detection
-    LOG_DEBUG(TRACE, "re-enabling charging");
-    chargingDisabledTimestamp_ = 0;
-    g_batteryState = BATTERY_STATE_UNKNOWN;
-    initDefault(false);
+void PowerManager::deduceBatteryStateChargeDisabled() {
+  DBG_PWR("deduceBatteryStateChargeDisabled()");
+  // XXX: This is a dirty hack. Using the VCELL to deduce the battery state
+  // is based on the experience that when charging is disabled, the VCELL can
+  // correctly be measured.
+  // When charging is disabled, the VCELL can reliably reflect the voltage on VBAT
+  FuelGauge fuel(true);
+  float vCell = fuel.getVCell();
+  LOG_DEBUG(TRACE, "VCell: %dmV", (uint16_t)(vCell * 1000));
+
+  // Make an assumption
+  if (g_batteryState == BATTERY_STATE_UNKNOWN) {
+    DBG_PWR("Battery state unknown, make an assumption:");
+    if (vCell < BATTERY_CONNECTED_VCELL_THRESHOLD) {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+    } else {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_NOT_CHARGING);
+    }
+    return;
+  } else if (g_batteryState == BATTERY_STATE_DISCONNECTED) {
+    if (vCell > BATTERY_CONNECTED_VCELL_THRESHOLD) {
+      if (vcellDebounceCount_ == 0) {
+        // It's probably that the battery is attached, request to shorten the monitor period
+        battMonitorPeriod_ = BATTERY_STATE_CHANGE_CHECK_PERIOD;
+        DBG_PWR("Battery might be attached");
+      }
+      vcellDebounceCount_++;
+      if (vcellDebounceCount_ >= BATTERY_VCELL_DEBOUNCE_COUNT) {
+        goto connected;
+      }
+    } else {
+      confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+    }
+    return;
+  }
+
+connected:
+  PMIC power(true);
+  const uint8_t status = power.getSystemStatus();
+  if (status & 0x08) {
+    DBG_PWR("In DPM mode");
+    // It's in DPM mode, the battery is discharging
+    confirmBatteryState(g_batteryState, BATTERY_STATE_DISCHARGING);
+  } else if (vCell < BATTERY_CONNECTED_VCELL_THRESHOLD) {
+    DBG_PWR("vCell < BATTERY_CONNECTED_VCELL_THRESHOLD");
+    // If battery falls below 2.0v, it's probably detached
+    confirmBatteryState(g_batteryState, BATTERY_STATE_DISCONNECTED);
+  } else {
+    // Since charging is disabled, it must be in the not charging shate
+    confirmBatteryState(g_batteryState, BATTERY_STATE_NOT_CHARGING);
```

Comment:
Do we additionally need to check 'power good' here or is DPM state enough?, same in `deduceBatteryStateChargeEnabled`

Reply:
It checks the DPM state a few lines above, but yes, we need to check the "power good" as well.